### ISSUE TITLE
[QUERY] Does the new library for Service Bus (Azure.Messaging.ServiceBus) support creating SAS Signatures? (Also troubleshooting recommendations)

### ISSUE THREAD

**Issue by mattmazzola** [Azure/azure-sdk-for-net#28929]
### Library name and version

Azure.Messaging.ServiceBus 7.8.1

### Query/Question

I have two questions:

1. I wanted to know if the new library for Service Bus (Azure.Messaging.ServiceBus) will support creating SAS Signatures?

I do not see the feature listed in the Migration guide and I have checked the API of the AdminClient and regular Client and assume it does not.

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/MigrationGuide.md#guide-for-migrating-to-azuremessagingservicebus-from-microsoftazureservicebus

I have been using the techniques here:
1. Manual Generation:
    https://docs.microsoft.com/en-us/rest/api/eventhub/generate-sas-token#c

1. Using TokenProvider
    ```csharp
    var provider = (SharedAccessSignatureTokenProvider)TokenProvider.CreateSharedAccessSignatureTokenProvider(keyName, key, validityDuration, tokenScope);
   var sasToken = await provider.GetTokenAsync(endpoint, validityDuration);
    ```
    https://github.com/Azure/azure-service-bus/blob/31a0018c122fb39e7016a25706023b45d9622374/samples/DotNet/Microsoft.Azure.ServiceBus/ManagingEntities/SASAuthorizationRule/Program.cs#L70-L72
    
   This latter technique seems more reliable since there is more use of tested library code instead of user code, but this requires taking dependency on BOTH `Microsoft.Azure.ServiceBus` and `Azure.Messaging.ServiceBus`. It was my understanding `Azure.Messaging.ServiceBus` was intended to replace `Microsoft.Azure.ServiceBus`.

1. What is the recommended way to troubleshoot issues with SAS Tokens?
  Such as: Is there any way to test SAS token generation and usage through the portal to verify configuration?

  I have tried _both_ techniques above they generate signatures, but when I try to use them to connect to ServiceBus to send or receive messages there is exception
   ```
   System.UnauthorizedAccessException: Put token failed. status-code: 401, status-description: InvalidSignature: The token has an invalid signature.
   ```
   I have tried many variations of configuration with resource URI path, entity path, etc, and am unsure how to continue. Any tips would be appreciated.

### Environment

.NET SDK (reflecting any global.json):
 Version:   6.0.300
 Commit:    8473146e7d

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22610
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.300\

Host (useful for support):
  Version: 6.0.5
  Commit:  70ae3df4a6

**Comment by mattmazzola** [comment]
I also wanted to note that there seems to be mix in the documentation. Some samples use the new Azure.Messaging.ServiceBus and others use Microsoft.Azure.ServiceBus. 

For example, in the code I am using to test SAS signatures I am taking the connection from the token provider 
```
var sasToken = await sasTokenProvider.GetTokenAsync(tokenAudience, validityDuration);
var serviceBusConnectionStringBuilder = new ServiceBusConnectionStringBuilder(endpoint, entityPath, sasToken.TokenValue);
var namespaceConnectionStringFromSas = serviceBusConnectionStringBuilder.GetNamespaceConnectionString();
```

and directly trying to create a client from it
```
var client = new ServiceBusClient(connectionStringOrSasToken);
var sender = client.CreateSender(topicName);
```

However, the documentation for Authentication using SAS here https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-sas#c is using the older method and library

```
var csStringBuilder = new ServiceBusConnectionStringBuilder(connectionStringOrSasToken);
var connection = new ServiceBusConnection(csStringBuilder);
var sender = new MessageSender(connection, topicName);
```

I think this documentation could be updated to use the new library.

**Comment by jsquire** [comment]
Hi @mattmazzola.    Thank you for reaching out and we regret that you're experiencing difficulties.   To answer your questions:

> I wanted to know if the new library for Service Bus (Azure.Messaging.ServiceBus) will support creating SAS Signatures?

No; the client library does not generate SAS tokens nor connection strings.  As a general rule, the current generation of Azure SDKs are intended to support connection strings and credentials only in the form returned by the Azure Portal.   In the case of Service Bus, I think a case could be made for supporting SAS generation, since there is no Portal support.   However, it's not a feature that we've received many requests for.  To my knowledge, the [documentation](https://docs.microsoft.com/azure/service-bus-messaging/service-bus-sas#c) that you've referenced is the recommended approach for generating them.   

> What is the recommended way to troubleshoot issues with SAS Tokens?

For troubleshooting, I'll have to defer to the service team.   The client only uses the provided credential and passes it to Service Bus.  @EldertGrootenboer may have insight.

> For example, in the code I am using to test SAS signatures I am taking the connection from the token provider and directly trying to create a client from it

Using a client with a SAS token would look a bit different than the pattern you're using:

```csharp
var fullyQualifiedNamespace = "<< NAMESPACE (likely similar to {your-namespace}.servicebus.windows.net) >>";
var credential = new AzureSasCredential("<< SHARED ACCESS KEY STRING >>");
var client = new ServiceBusClient(fullyQualifiedNamespace , credential);
```
  The advantage to this approach is that the SAS token can be updated without creating a new client, using the [AzureSasCredential.Update](https://docs.microsoft.com//dotnet/api/azure.azuresascredential.update?view=azure-dotnet#azure-azuresascredential-update(system-string)) method.   
_(NOTE: There is a gap in the migration guide, currently, for moving from the token provider.   That's being tracked by #25787)_

> However, the documentation for Authentication using SAS here https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-sas#c is using the older method and library

I went to open an issue for the MS Docs content team to update that page, but I'm not seeing the code snippet that you've included.   Was that from a different page?

**Comment by jsquire** [comment]
@EldertGrootenboer:   Do you have any insight on recommendations for testing SAS keys?    From the issue:

```
What is the recommended way to troubleshoot issues with SAS Tokens?
Such as: Is there any way to test SAS token generation and usage through the portal to verify configuration?

I have tried both techniques above they generate signatures, but when I try to use them to connect to ServiceBus to send or receive messages there is exception

System.UnauthorizedAccessException: Put token failed. status-code: 401, status-description: InvalidSignature: The token has an invalid signature.
I have tried many variations of configuration with resource URI path, entity path, etc, and am unsure how to continue. Any tips would be appreciated.
```

**Comment by mattmazzola** [comment]
> the current generation of Azure SDKs are intended to support connection strings and keys only in the form returned by the Azure Portal.

Does this mean I should expect code like below which attempts to create `ServiceBusClient` using a connection string with SAS token to fail? (Because this is a NOT a connection string that could be retrieved from the portal.)

```csharp
var client = new ServiceBusClient("Endpoint=sb://plaiground-servicebus.servicebus.windows.net/;SharedAccessSignature=SharedAccessSignature sr=sb%3a%2f%2f<removed>-servicebus.servicebus.windows.net%2f&sig=D8qDhmgD3L%2fk<removed>JcE%2fOk9vhsITn2pc%3d&se=1653345169&skn=key1;");
```

I thought any connection string would work regardless if it was using SharedAccessKey or SharedAccessSignature.

> Using a client with a SAS token would look a bit different than the pattern you're using

Thanks for the suggestion. I have tried using this technique which direction constructs the client and avoids need for full connection string which may have avoids some of the code issues, but I still receive the same error about PUT token failed, invalid signature. I expected this since the information was (token) was the same, but wanted to try anyways since I am out of options.

> I'm not seeing the code snippet that you've included. Was that from a different page?

I see it was mixing of separate snippets and slight confusion on my part. That snipped I used above was from my code that I'm using to test the various ways to generate SAS token from Key and various ways to create a client from that token. However, I see the documentation I liked which is actually showing "how to send the SAS token with AMQP protocol using the AMQP.NET Lite library." so it's not EITHER library.

This is code which connects to Service Bus using SAS Token. It seems awkward that it specifies the connection string AND the sas token separately when usually the connection string already contains credentials.

https://github.com/Azure/azure-service-bus/blob/31a0018c122fb39e7016a25706023b45d9622374/samples/DotNet/Microsoft.Azure.ServiceBus/ManagingEntities/SASAuthorizationRule/Program.cs#L77-L81

I suppose in worst case if I cannot get SAS connection working I can manually create Shared Access Policy for each subscription so it has very narrow scope.

**Comment by jsquire** [comment]
> Does this mean I should expect code like below which attempts to create ServiceBusClient using a connection string with SAS token to fail? (Because this is a NOT a connection string that could be retrieved from the portal.)

Generally, that is a true statement.  In this case, there is an exception for `SharedAccessSignature`; it will continue to be work and will be supported.  The other custom tokens introduced by the legacy packages for things like setting configuration options are no longer recognized.  That said, our recommendation for working with SAS tokens is to use `AzureSasCredential`.

**Comment by mattmazzola** [comment]
Thanks for the help Jesse!
I have tried your suggested code in place of what I had and I get the same error 😞
I think it is very similar to the code I had but is using the "amqps".

I have written a CLI utility that uses all 3 methods of generating SAS to compare the output signatures.
and another that goes through sequence:
1. Generating Shared Access Signature (using your code above)
1. Create a ServiceBusClient with FQN and AzureSasCredential from SAS in step 1
1. Attempt to send messages

> Put token failed. status-code: 401, status-description: InvalidSignature: The token has an invalid signature.

I feel like I've exhaustively tried everything 😅.  

Also, based on this
> not a feature that we've received many requests for
I thought it would help to give some more context on our what we are trying to do.

I think in most cases the producers and consumers of ServiceBus are under your control and can be trusted with a key, but we expect to have a dynamic number of consumers who are external. We wanted to be able to dynamically create subscription for the user and a SAS for them to use so we don't have worry about key distribution.

**Comment by jsquire** [comment]
Interesting.  Crazy question.... do you happen to have shared keys at the namespace level and entity level that have the same exact name?

**Comment by mattmazzola** [comment]
> do you happen to have shared keys at the namespace level and entity level that have the same exact name?

No, I have the default `RootManageSharedAccessKey` at namespace level and that is the only policy.

I assume somehow during preparation for the making the video I changed some code, but I looked and could not find it.  I even changed the configuration back to use my Original service bus instance, where the SAS were failing with invalid token, and they started working too 🤯.  I was suspicious that somehow I was not actually using SAS and somehow still using the normal key, but I don't think so either.  

Anyways, I have already spent too much time on this 😅 I may never know what happened but at least it's working.
